stone/frontend/ir_generator.py

```diff
@@ -319,6 +328,73 @@ def _create_type(self, env, item):
         env[item.name] = api_type
         return api_type
 
+    def _merge_patches(self):
+        """Injects object patches into their original object definitions."""
+        for patched_item, patched_namespace in self._patch_data_by_canonical_name.values():
+            patched_item_base_name = self._get_base_name(patched_item.name, patched_namespace.name)
+            if patched_item_base_name not in self._item_by_canonical_name:
+                raise InvalidSpec('Patch {} must correspond to a pre-existing data_type.'.format(
+                    quote(patched_item.name)), patched_item.lineno, patched_item.path)
+
+            existing_item = self._item_by_canonical_name[patched_item_base_name]
+
+            self._check_patch_type_mismatch(patched_item, existing_item)
+
+            if (isinstance(patched_item, AstStructPatch)
+                    or isinstance(patched_item, AstUnionPatch)):
+                self._check_field_names_unique(existing_item, patched_item)
+                existing_item.fields += patched_item.fields
+                self._inject_patched_examples(existing_item, patched_item)
+            else:
+                raise AssertionError('Unknown Patch Object Type {}'.format(
+                    patched_item.__class__.__name__))
+
+    def _check_patch_type_mismatch(self, patched_item, existing_item):
+        """Enforces that each patch has a corresponding, already-defined data type."""
+        def raise_mismatch_error(patched_item, existing_item, data_type_name):
+            error_msg = ('Type mismatch. Patch {} corresponds to pre-existing '
+                'data_type {} ({}:{}) that has type other than {}.')
+            raise InvalidSpec(error_msg.format(
+                quote(patched_item.name),
+                quote(existing_item.name),
+                existing_item.path,
+                existing_item.lineno,
+                quote(data_type_name)), patched_item.lineno, patched_item.path)
+
+        if isinstance(patched_item, AstStructPatch):
+            if not isinstance(existing_item, AstStructDef):
+                raise_mismatch_error(patched_item, existing_item, 'struct')
+        elif isinstance(patched_item, AstUnionPatch):
```

Comment:
Enforce that the original union data type and union patched data type have a matching close policy. Either both closed or open.

Reply:
Good catch, I added a check / corresponding test.